### ISSUE TITLE
Improving thread safety of FMResultSet

### ISSUE THREAD

**Issue by benasher44** [ccgus/fmdb#650]
Hello! We've been noticing some crashes (although relatively few) in production when closing an `FMResultSet`. The crash happens when `FMResultSet` contacts its `parentDB` to remove itself from the set of open results sets. We believe the crashing is occurring because it's not safe to mutate a container from multiple threads. That said, I believe it's safe to close and otherwise use sqlite result sets from the same db connection on different threads (lets assume each result set gets its own thread) when sqlite is in serialized mode, which is the [default](https://www.sqlite.org/threadsafe.html).

What do you all think would be the best course of action here? Would you all agree that this is something that should be fixed, and if so, do you think it'd make sense to add an internal queue/lock to synchronize the mutable containers internal to `FMDatabase`?

Thanks for your consideration!

**Comment by robertmryan** [comment]
I assume you are not using `FMDatabaseQueue`?

**Comment by groue** [comment]
@benasher44 What is the purpose adding internal queues/locks to FMDatabase, when the queuing job is already provided by FMDatabaseQueue?

In other words, did you consider replacing your pool of FMDatabase for reads with a pool of read-only FMDatabaseQueue? This would greatly help your implementation.

And you'd even be able to use FMDatabase features that are not thread-safe, like statement caching. Generally speaking, FMDatabase is not thread-safe. None of its [ivars](https://github.com/ccgus/fmdb/blob/master/src/fmdb/FMDatabase.m#L11) are protected.

**Comment by benasher44** [comment]
The purpose would be to allow using a single `FMResultSet` on its own thread. I agree that making `FMDatabase` and `FMResultSet` fully thread-safe is a non-goal because that's what `FMDatabaseQueue` is for. However, to properly use an `FMResultSet` in a thread-safe way today, you have to extract every row from the result set in the context of an `FMDatabaseQueue`, which means you'll spend a lot of time blocking the queue while you're extracting all of your result rows. What I'd like to be able to do is get a result set and then extract all of its rows on another thread, freeing the queue. This isn't possible today because closing the result set from another thread causes will touch `FMDatabase` in a way that isn't thread safe. Apologies for not making this clearer sooner! Does that make sense?

**Comment by benasher44** [comment]
So yes let’s say tomorrow I move forward with a pool of `FMDatabaseQueue` instances. I’m still restricted to using any `FMResultSet`s I get on the `FMDatabaseQueue`’s queue, despite intending to use the result set itself on its own thread/queue.

**Comment by benasher44** [comment]
So to further clarify, that’s the restriction I’d like to lift. `FMResultSet` would still not be thread safe, but it would no longer be restricted to use only on its owning parentDB’s queue.

**Comment by groue** [comment]
> The purpose would be to allow using a single FMResultSet on its own thread.
I agree that making FMDatabase and FMResultSet fully thread-safe is a non-goal because that's what FMDatabaseQueue is for. However, to properly use an FMResultSet in a thread-safe way today, you have to extract every row from the result set in the context of an FMDatabaseQueue, which means you'll spend a lot of time blocking the queue while you're extracting all of your result rows.

Yes, we sometimes want to release queues as early as possible. But not always:

If the queue is a reader queue, then you don't mind blocking it since other threads will simply use another reader queue from the pool.

If the queue is the writer queue, then yes indeed, you don't want to block the writer queue longer than necessary. Preciselly speaking, when you write, then read, you don't want the read to block the writer queue.

> What I'd like to be able to do is get a result set and then extract all of its rows on another thread, freeing the queue. This isn't possible today because closing the result set from another thread will touch FMDatabase in a way that isn't thread safe.

Yes. So don't use the result set from any other thread.

> Apologies for not making this clearer sooner! Does that make sense?

A lot of sense ;-)

If you use the WAL mode, then here is a technique that acheives your goal, while keeping connections and result sets under strict control of database queues.

Remember: we just want to avoid blocking the writer queue longer than necessary when we want to read freshly updated requests. All other cases are already handled by the readers pool.

Here it is: after you have committed changes and want to release the writer queue, spawn a reader from the riter queue. Put the reader in snapshot isolation (see below). Only then, you can release the writer queue.

You get a reader can will read exactly what have been last committed in the database, and yet the writer queue is no longer blocked. Close the deferred transaction to leave snapshot isolation.

> I’m still restricted to using any FMResultSets I get on the FMDatabaseQueue’s queue, despite intending to use the result set itself on its own thread/queue.

Yes. And it is no longer something that bugs you, because you have acheived the precise scheduling you need with plain old database queues. Even if FMDatabase and FMResultSet are still unsafe.

**What is snapshot isolation?**

https://www.sqlite.org/isolation.html

> In WAL mode, SQLite exhibits "snapshot isolation". When a read
> transaction starts, that reader continues to see an unchanging
> "snapshot" of the database file as it existed at the moment in time
> when the read transaction started. Any write transactions that
> commit while the read transaction is active are still invisible to
> the read transaction, because the reader is seeing a snapshot of
> database file from a prior moment in time.

http://www.sqlite.org/lang_transaction.html

> Deferred transaction means that no locks are acquired on the
> database until the database is first accessed. [...] Locks are not
> acquired until the first read or write operation. [...] Because the
> acquisition of locks is deferred until they are needed, it is
> possible that another thread or process could create a separate
> transaction and write to the database after the BEGIN on the
> current thread has executed.

So basically, to make sure that the spawned reader reads exactly what you want, you open a deferred transaction and read something (anything), and only then you release the writer queue.

**Comment by benasher44** [comment]
Sure that technique makes sense, and I appreciate the thorough explanation there. SQLite details aside and from a pure object resource perspective, once you get an `FMResultSet`, the work of the `FMDatabase` is done up until the point when you have to close the result set, at which point some state has to be updated internally to the `FMDatabase` instance. Unfortunately due to that constraint, it’s not safe to put the `FMDatabase` instance back into the pool until that point. For most people, I don’t think they’ll fully realize 
 that constraint until they start running into these kinds of crashes. At most, I’d like to add locking to allow putting the `FMDatabase` back into the pool sooner. At least, I’d like to document this behavior (that close should be called on an `FMDatabaseQueue`’s queue). If folks feel weary of adding such locking, I’m happy to just document this for now.

**Comment by benasher44** [comment]
> Yes, we sometimes want to release queues as early as possible. But not always:
> 
> If the queue is a reader queue, then you don't mind blocking it since other threads will simply use another reader queue from the pool.

Our app is particularly resource constrained, and it’s common to be processing 1000’s of objects from multiple threads at any given time. Unfortunately for us, holding onto `FMDatabaseQueue` instances longer than necessary for readers will likely lead to us needing to increase our pool size, which we’d like to avoid.

**Comment by groue** [comment]
Thanks for this detailed context. Your last comment shows that database queues are not enough for your app, and that you will not use a pool of reader queues.

**Comment by ccgus** [comment]
Is there any reason you can't have a single FMDatabase instance for each FMResult set?

**Comment by benasher44** [comment]
I suppose having more open read connections shouldn't be an issue. I'll give that a try and update this thread if I run into issues there. Otherwise, I'll close it :)

**Comment by benasher44** [comment]
Thanks for hearing me out!

**Comment by benasher44** [comment]
The minor tweaks we made around this turned out to be pretty reasonable in production. Going to close. Thanks again!
